Graal Native Build missing ReflectionRegistrations in io.fabric8:kubernetes-model-core:5.12.3

### Describe the bug

Hi

The fix should be fast. I get following exception when running my native app do to missing ReflectionRegistrations:
Caused by: java.lang.NoSuchMethodException: io.fabric8.kubernetes.api.model.PersistentVolumeClaimBuilder.<init>(io.fabric8.kubernetes.api.model.PersistentVolumeClaim)
        at java.lang.Class.getConstructor0(DynamicHub.java:3585)
        at java.lang.Class.getDeclaredConstructor(DynamicHub.java:2754)
        at io.fabric8.kubernetes.client.ResourceHandlerImpl.edit(ResourceHandlerImpl.java:62)
        ... 38 more

I solved it by adding this class to my code:

import io.fabric8.kubernetes.api.model.ConfigMapBuilder;
import io.fabric8.kubernetes.api.model.PersistentVolumeClaimBuilder;
import io.quarkus.runtime.annotations.RegisterForReflection;

/**
 * Temporary registration of external classes which should be fixed by the library in the future.
 * This patch is neede for quarkus native build
 * TODO remove as soon Openshif-Client library patched this bug
 */
@RegisterForReflection(targets = {PersistentVolumeClaimBuilder.class, ConfigMapBuilder.class})
public class RegisterReflections {

}

But it should be fixed in the library.


### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

I think the full Stacktrace is not even necessary, cause the fix is so easy as described above. The same behavior happened for io.fabric8.kubernetes.api.model.ConfigMapBuilder but here the full Stacktrace:

2022-08-23 16:35:26,321 ERROR [ch.sbb.est.tek.pip.ser.ope.OpenshiftTemplateService] (vert.x-eventloop-thread-4) Failed to modify and process template esta-tekton-pipeline-java-maven-build-run-template. Reason: An error has occurred.: io.fabric8.kubernetes.client.KubernetesClientException: An error has occurred.
        at io.fabric8.kubernetes.client.ResourceHandlerImpl.edit(ResourceHandlerImpl.java:65)
        at io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.acceptVisitors(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.java:245)
        at io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.inNamespace(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.java:129)
        at io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.inNamespace(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.java:63)
        at ch.sbb.esta.tekton.pipeline.service.openshift.OpenshiftTemplateService.lambda$processAndApplyTemplate$2(OpenshiftTemplateService.java:96)
        at java.util.ArrayList.forEach(ArrayList.java:1511)
        at ch.sbb.esta.tekton.pipeline.service.openshift.OpenshiftTemplateService.processAndApplyTemplate(OpenshiftTemplateService.java:92)
        at ch.sbb.esta.tekton.pipeline.service.openshift.OpenshiftTemplateService.processAndApplyWithTmpTemplate(OpenshiftTemplateService.java:73)
        at ch.sbb.esta.tekton.pipeline.service.openshift.OpenshiftTemplateService.modifyProcessAndApplyTemplate(OpenshiftTemplateService.java:65)
        at ch.sbb.esta.tekton.pipeline.service.openshift.OpenshiftTemplateService.modifyProcessAndApplyPipelineRunTemplate(OpenshiftTemplateService.java:50)
        at ch.sbb.esta.tekton.pipeline.service.openshift.OpenshiftTemplateService_ClientProxy.modifyProcessAndApplyPipelineRunTemplate(Unknown Source)
        at ch.sbb.esta.tekton.pipeline.service.pipelinestarter.PipelineExecutor.execPipelineRunFromTemplate(PipelineExecutor.java:68)
        at ch.sbb.esta.tekton.pipeline.service.pipelinestarter.PipelineExecutor.startPipeline(PipelineExecutor.java:52)
        at ch.sbb.esta.tekton.pipeline.service.pipelinestarter.PipelineExecutor_ClientProxy.startPipeline(Unknown Source)
        at ch.sbb.esta.tekton.pipeline.service.pipelinestarter.PipelineEventProcessor.startPipeline(PipelineEventProcessor.java:38)
        at ch.sbb.esta.tekton.pipeline.service.pipelinestarter.PipelineEventProcessor_ClientProxy.startPipeline(Unknown Source)
        at ch.sbb.esta.tekton.pipeline.service.pipelinestarter.PipelineEventProcessor_VertxInvoker_startPipeline_b2682422d03771df196866d5169809e873d6f3cf.invokeBean(Unknown Source)
        at io.quarkus.vertx.runtime.EventConsumerInvoker.invoke(EventConsumerInvoker.java:41)
        at io.quarkus.vertx.runtime.VertxRecorder$3$1.handle(VertxRecorder.java:130)
        at io.quarkus.vertx.runtime.VertxRecorder$3$1.handle(VertxRecorder.java:100)
        at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:264)
        at io.vertx.core.eventbus.impl.MessageConsumerImpl.dispatch(MessageConsumerImpl.java:177)
        at io.vertx.core.eventbus.impl.HandlerRegistration$InboundDeliveryContext.execute(HandlerRegistration.java:137)
        at io.vertx.core.eventbus.impl.DeliveryContextBase.next(DeliveryContextBase.java:72)
        at io.vertx.core.eventbus.impl.DeliveryContextBase.dispatch(DeliveryContextBase.java:43)
        at io.vertx.core.eventbus.impl.HandlerRegistration.dispatch(HandlerRegistration.java:98)
        at io.vertx.core.eventbus.impl.MessageConsumerImpl.deliver(MessageConsumerImpl.java:183)
        at io.vertx.core.eventbus.impl.MessageConsumerImpl.doReceive(MessageConsumerImpl.java:168)
        at io.vertx.core.eventbus.impl.HandlerRegistration.lambda$receive$0(HandlerRegistration.java:49)
        at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:503)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:833)
        at com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:705)
        at com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine(PosixPlatformThreads.java:202)
Caused by: java.lang.NoSuchMethodException: io.fabric8.kubernetes.api.model.PersistentVolumeClaimBuilder.<init>(io.fabric8.kubernetes.api.model.PersistentVolumeClaim)
        at java.lang.Class.getConstructor0(DynamicHub.java:3585)
        at java.lang.Class.getDeclaredConstructor(DynamicHub.java:2754)
        at io.fabric8.kubernetes.client.ResourceHandlerImpl.edit(ResourceHandlerImpl.java:62)
        ... 38 more

--- a/extensions/jackson/deployment/src/main/java/io/quarkus/jackson/deployment/IgnoreJsonDeserializeClassBuildItem.java
+++ b/extensions/jackson/deployment/src/main/java/io/quarkus/jackson/deployment/IgnoreJsonDeserializeClassBuildItem.java
@@ -1,5 +1,7 @@
 package io.quarkus.jackson.deployment;
 
+import java.util.List;
+
 import org.jboss.jandex.DotName;
 
 import io.quarkus.builder.item.MultiBuildItem;
@@ -10,13 +12,22 @@ import io.quarkus.builder.item.MultiBuildItem;
  */
 public final class IgnoreJsonDeserializeClassBuildItem extends MultiBuildItem {
 
-    private final DotName dotName;
+    private final List<DotName> dotNames;
 
     public IgnoreJsonDeserializeClassBuildItem(DotName dotName) {
-        this.dotName = dotName;
+        this.dotNames = List.of(dotName);
     }
 
+    public IgnoreJsonDeserializeClassBuildItem(List<DotName> dotNames) {
+        this.dotNames = dotNames;
+    }
+
+    @Deprecated(forRemoval = true)
     public DotName getDotName() {
-        return dotName;
+        return dotNames.size() > 0 ? dotNames.get(0) : null;
+    }
+
+    public List<DotName> getDotNames() {
+        return dotNames;
     }
 }

--- a/extensions/jackson/deployment/src/main/java/io/quarkus/jackson/deployment/JacksonProcessor.java
+++ b/extensions/jackson/deployment/src/main/java/io/quarkus/jackson/deployment/JacksonProcessor.java
@@ -123,7 +123,7 @@ public class JacksonProcessor {
 
         Set<DotName> ignoredDotNames = new HashSet<>();
         for (IgnoreJsonDeserializeClassBuildItem ignoreJsonDeserializeClassBuildItem : ignoreJsonDeserializeClassBuildItems) {
-            ignoredDotNames.add(ignoreJsonDeserializeClassBuildItem.getDotName());
+            ignoredDotNames.addAll(ignoreJsonDeserializeClassBuildItem.getDotNames());
         }
 
         // handle the various @JsonDeserialize cases

--- a/extensions/kubernetes-client/deployment/src/main/java/io/quarkus/kubernetes/client/deployment/KubernetesClientProcessor.java
+++ b/extensions/kubernetes-client/deployment/src/main/java/io/quarkus/kubernetes/client/deployment/KubernetesClientProcessor.java
@@ -52,6 +52,7 @@ public class KubernetesClientProcessor {
 
     private static final Predicate<DotName> IS_OKHTTP_CLASS = d -> d.toString().startsWith("okhttp3");
     private static final DotName JSON_FORMAT = DotName.createSimple(JsonFormat.class.getName());
+    private static final String[] EMPTY_STRINGS_ARRAY = new String[0];
 
     @BuildStep
     public void registerBeanProducers(BuildProducer<AdditionalBeanBuildItem> additionalBeanBuildItemBuildItem,
@@ -148,11 +149,13 @@ public class KubernetesClientProcessor {
                 });
         if (!withFieldsRegistration.isEmpty()) {
             reflectiveClasses.produce(ReflectiveClassBuildItem
-                    .builder(withFieldsRegistration.toArray(new String[0])).weak(true).methods(true).fields(true).build());
+                    .builder(withFieldsRegistration.toArray(EMPTY_STRINGS_ARRAY)).weak(true).methods(true).fields(true)
+                    .build());
         }
         if (!withoutFieldsRegistration.isEmpty()) {
             reflectiveClasses.produce(ReflectiveClassBuildItem
-                    .builder(withoutFieldsRegistration.toArray(new String[0])).weak(true).methods(true).fields(false).build());
+                    .builder(withoutFieldsRegistration.toArray(EMPTY_STRINGS_ARRAY)).weak(true).methods(true).fields(false)
+                    .build());
         }
 
         // we also ignore some classes that are annotated with @JsonDeserialize that would force the registration of the entire model

--- a/extensions/kubernetes-client/deployment/src/main/java/io/quarkus/kubernetes/client/deployment/KubernetesClientProcessor.java
+++ b/extensions/kubernetes-client/deployment/src/main/java/io/quarkus/kubernetes/client/deployment/KubernetesClientProcessor.java
@@ -46,6 +46,11 @@ public class KubernetesClientProcessor {
             .createSimple("io.fabric8.kubernetes.client.informers.ResourceEventHandler");
     private static final DotName KUBERNETES_RESOURCE = DotName
             .createSimple("io.fabric8.kubernetes.api.model.KubernetesResource");
+    private static final DotName KUBERNETES_RESOURCE_LIST = DotName
+            .createSimple("io.fabric8.kubernetes.api.model.KubernetesResourceList");
+
+    private static final DotName VISITABLE_BUILDER = DotName
+            .createSimple("io.fabric8.kubernetes.api.builder.VisitableBuilder");
     private static final DotName CUSTOM_RESOURCE = DotName.createSimple("io.fabric8.kubernetes.client.CustomResource");
 
     private static final Logger log = Logger.getLogger(KubernetesClientProcessor.class.getName());
@@ -117,36 +122,28 @@ public class KubernetesClientProcessor {
 
         Collection<ClassInfo> kubernetesResourceImpls = combinedIndexBuildItem.getIndex()
                 .getAllKnownImplementors(KUBERNETES_RESOURCE);
+        Collection<ClassInfo> kubernetesResourceListImpls = combinedIndexBuildItem.getIndex()
+                .getAllKnownImplementors(KUBERNETES_RESOURCE_LIST);
+        Collection<ClassInfo> visitableBuilderImpls = combinedIndexBuildItem.getIndex()
+                .getAllKnownImplementors(VISITABLE_BUILDER);
+
         // default sizes determined experimentally - these are only set in order to prevent continuous expansion of the array list
-        List<String> withoutFieldsRegistration = new ArrayList<>(kubernetesResourceImpls.size());
+        List<String> withoutFieldsRegistration = new ArrayList<>(
+                kubernetesResourceImpls.size() + kubernetesResourceListImpls.size());
         List<String> withFieldsRegistration = new ArrayList<>(2);
-        kubernetesResourceImpls
-                .stream()
-                .peek(c -> {
-                    // we need to make sure that the Jackson extension does not try to fully register the model classes
-                    // since we are going to register them weakly
-                    ignoredJsonDeserializationClasses.produce(new IgnoreJsonDeserializeClassBuildItem(c.name()));
-                })
-                .filter(c -> !watchedClasses.contains(c.name()))
-                .map(c -> {
-                    boolean registerFields = false;
-                    List<AnnotationInstance> jsonFormatInstances = c.annotationsMap().get(JSON_FORMAT);
-                    if (jsonFormatInstances != null) {
-                        for (AnnotationInstance jsonFormatInstance : jsonFormatInstances) {
-                            if (jsonFormatInstance.target().kind() == AnnotationTarget.Kind.FIELD) {
-                                registerFields = true;
-                                break;
-                            }
-                        }
-                    }
-                    return new AbstractMap.SimpleEntry<>(c.name(), registerFields);
-                }).forEach(e -> {
-                    if (e.getValue()) {
-                        withFieldsRegistration.add(e.getKey().toString());
-                    } else {
-                        withoutFieldsRegistration.add(e.getKey().toString());
-                    }
-                });
+        List<DotName> ignoreJsonDeserialization = new ArrayList<>(
+                kubernetesResourceImpls.size() + kubernetesResourceListImpls.size());
+
+        populateReflectionRegistrationLists(kubernetesResourceImpls, watchedClasses, ignoreJsonDeserialization,
+                withoutFieldsRegistration,
+                withFieldsRegistration);
+        populateReflectionRegistrationLists(kubernetesResourceListImpls, watchedClasses, ignoreJsonDeserialization,
+                withoutFieldsRegistration,
+                withFieldsRegistration);
+        populateReflectionRegistrationLists(visitableBuilderImpls, watchedClasses, ignoreJsonDeserialization,
+                withoutFieldsRegistration,
+                withFieldsRegistration);
+
         if (!withFieldsRegistration.isEmpty()) {
             reflectiveClasses.produce(ReflectiveClassBuildItem
                     .builder(withFieldsRegistration.toArray(EMPTY_STRINGS_ARRAY)).weak(true).methods(true).fields(true)
@@ -157,6 +154,8 @@ public class KubernetesClientProcessor {
                     .builder(withoutFieldsRegistration.toArray(EMPTY_STRINGS_ARRAY)).weak(true).methods(true).fields(false)
                     .build());
         }
+
+        ignoredJsonDeserializationClasses.produce(new IgnoreJsonDeserializeClassBuildItem(ignoreJsonDeserialization));
 
         // we also ignore some classes that are annotated with @JsonDeserialize that would force the registration of the entire model
         ignoredJsonDeserializationClasses.produce(
@@ -204,6 +203,40 @@ public class KubernetesClientProcessor {
 
         // Enable SSL support by default
         sslNativeSupport.produce(new ExtensionSslNativeSupportBuildItem(Feature.KUBERNETES_CLIENT));
+    }
+
+    private static void populateReflectionRegistrationLists(Collection<ClassInfo> kubernetesResourceImpls,
+            Set<DotName> watchedClasses,
+            List<DotName> ignoredJsonDeserializationClasses,
+            List<String> withoutFieldsRegistration,
+            List<String> withFieldsRegistration) {
+        kubernetesResourceImpls
+                .stream()
+                .peek(c -> {
+                    // we need to make sure that the Jackson extension does not try to fully register the model classes
+                    // since we are going to register them weakly
+                    ignoredJsonDeserializationClasses.add(c.name());
+                })
+                .filter(c -> !watchedClasses.contains(c.name()))
+                .map(c -> {
+                    boolean registerFields = false;
+                    List<AnnotationInstance> jsonFormatInstances = c.annotationsMap().get(JSON_FORMAT);
+                    if (jsonFormatInstances != null) {
+                        for (AnnotationInstance jsonFormatInstance : jsonFormatInstances) {
+                            if (jsonFormatInstance.target().kind() == AnnotationTarget.Kind.FIELD) {
+                                registerFields = true;
+                                break;
+                            }
+                        }
+                    }
+                    return new AbstractMap.SimpleEntry<>(c.name(), registerFields);
+                }).forEach(e -> {
+                    if (e.getValue()) {
+                        withFieldsRegistration.add(e.getKey().toString());
+                    } else {
+                        withoutFieldsRegistration.add(e.getKey().toString());
+                    }
+                });
     }
 
     private void findWatchedClasses(final DotName implementedOrExtendedClass, final ApplicationIndexBuildItem applicationIndex,

